src/oc/fileio.cpp

```diff
@@ -473,61 +486,90 @@ void hoc_sprint1(char** ppbuf, int argn) { /* convert args to right type for con
     *ppbuf = hs->buf;
 }
 
-#if defined(WIN32)
-static FILE* oc_popen(char const* const cmd, char const* const type) {
-    FILE* fp;
-    char buf[1024];
-    assert(strlen(cmd) + 20 < 1024);
-    Sprintf(buf, "sh %s > hocload.tmp", cmd);
-    if (system(buf) != 0) {
-        return (FILE*) 0;
-    } else if ((fp = fopen("hocload.tmp", "r")) == (FILE*) 0) {
-        return (FILE*) 0;
-    } else {
-        return fp;
+
+// Split a string containing an env variable into multiple paths (OS-specific)
+// and return a container with the results
+static auto split_paths(const std::string& input) {
+    std::vector<std::string> result;
+    std::size_t start = 0;
+    std::size_t end = 0;
+
+    while ((end = input.find(os_pathsep, start)) != std::string::npos) {
+        if (end > start) {
+            result.push_back(input.substr(start, end - start));
+        }
+        start = end + 1;
+    }
+
+    // Add the last segment if it's non-empty
+    if (start < input.size()) {
+        result.push_back(input.substr(start));
     }
+
+    return result;
 }
-static void oc_pclose(FILE* fp) {
-    fclose(fp);
-    unlink("hocload.tmp");
+
+// return the default search paths for loading files
+static auto default_search_paths() {
+    auto result = std::vector<std::string>({"."});
+
+    // insert hoc paths (if any)
+    auto hoc_library_path = getenv("HOC_LIBRARY_PATH");
+    if (hoc_library_path) {
+        auto paths = split_paths(std::string(hoc_library_path));
+        std::copy(begin(paths), end(paths), back_inserter(result));
+    }
+
+    // insert home path
+    auto home_path = fmt::format("{}/lib/hoc", std::string(neuron_home));
+    result.push_back(home_path);
+
+    return result;
+}
+
+// look for regex ``pattern`` in ``paths`` (non-recursively);
+// if there is no match, return empty value
+// if there is a match, return the path of the first file containing the match
+static std::optional<std::string> search_hoc_files_regex(const std::regex& pattern,
+                                                         const std::vector<std::string>& paths) {
+    namespace fs = std::filesystem;
+    for (const auto& path: paths) {
+        for (const auto& entry: fs::directory_iterator(path)) {
```

Comment:
Not sure if this returns the same ordering as `egrep '^func|^proc|^begintemplate' $p/*.oc $p/*.hoc`.

Reply:
The entries may not be sorted at all, whereas the shell bit you point out first goes through all `.oc` files, followed by all `.hoc`.

Since only the first match is used, would one expect multiple definitions of the same entity?

Reply:
> The entries may not be sorted at all

Indeed! I changed the implementation so it's actually sorted, it should even respect the locale now since that's what the POSIX `glob` does (had a whole rant ready about locales, but it turns out it's surprisingly simple to use, at least for this usecase :)

> Since only the first match is used, would one expect multiple definitions of the same entity?

I'd expect it to throw an error in case of multiple definitions, or provide some mechanism to disambiguate, but I don't think there is anything of the sort, so I just sought to replicate the existing behavior.